src/compiler/crystal/codegen/debug.cr

```diff
@@ -38,8 +38,14 @@ module Crystal
       di_builder.create_subroutine_type(nil, int1)
     end
 
-    def debug_type_cache
-      @debug_types ||= {} of Type => LibLLVMExt::Metadata?
+    def debug_type_cache(llvm_module = @llvm_mod)
+      # We must cache debug types per module so metadata of a type
+      # from one module isn't incorrectly used in another module.
+      debug_types_per_module =
+        @debug_types_per_module ||=
+          {} of LLVM::Module => Hash(Type, LibLLVMExt::Metadata?)
+
+      debug_types_per_module[llvm_module] ||= {} of Type => LibLLVMExt::Metadata?
```

Comment:
There is no invocation to `debug_type_cache` with arguments.
Why not use `@llvm_mod` directly instead of `llvm_module` as the index to access debug_types_per_module?

Reply:
Good idea! Please send a PR (I'll be busy today, water leak under my apartment :/)